### PR TITLE
Consider empty-unreachable otherwise branches in MatchBranchSimplification

### LINKED ISSUE #131219: [MIR-opt] `matches!(x, Some(_))` should simplify to `Eq(Discriminant, Const)`
Repro: https://rust.godbolt.org/z/4dvWdqcn5

Today, this check
```rust
pub fn my_is_eq<T>(x: &std::cmp::Ordering) -> bool { matches!(x, std::cmp::Ordering::Equal)}
```
simplifies as expected down to just
```rust
    bb0: {
        _2 = discriminant((*_1));
        _0 = Eq(copy _2, const 0_i8);
        return;
    }
```

Checking for `Some` in an `Option`, however,
```rust
pub fn my_is_some<T>(x: &Option<T>) -> bool { matches!(x, Some(_)) }
```
doesn't and ends up taking a whole *five* basic blocks
```rust
    bb0: {
        _2 = discriminant((*_1));
        switchInt(move _2) -> [1: bb2, 0: bb1, otherwise: bb4];
    }

    bb1: {
        _0 = const false;
        goto -> bb3;
    }

    bb2: {
        _0 = const true;
        goto -> bb3;
    }

    bb3: {
        return;
    }

    bb4: {
        unreachable;
    }
```

That's wasteful, and shouldn't happen.

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #131219

### CODE REVIEW

**Comment by clubby789** [comment]
Hmm, while I think this change is useful (given the improvement in another test), while checking the diffs I noticed that (on current master) the motivating example emits
```rs
fn my_is_some(_1: &Option<T>) -> bool {
    debug bar => _1;
    let mut _0: bool;
    let mut _2: isize;

    bb0: {
        _2 = discriminant((*_1));
        switchInt(move _2) -> [1: bb2, otherwise: bb1];   // two-branch SwitchInt, not 3
    }

    bb1: {
        _0 = const false;
        goto -> bb3;
    }

    bb2: {
        _0 = const true;
        goto -> bb3;
    }

    bb3: {
        return;
    }
}
```
And is therefore simplified by `MatchBranchSimplification` to the `Eq` version

File: tests/mir-opt/matches_reduce_branches.my_is_some.MatchBranchSimplification.diff [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+-     }
+- 
+-     bb3: {
+-         unreachable;
+-     }
+- 
+-     bb4: {
++         StorageLive(_3);
++         _3 = copy _2;
++         _0 = Ne(copy _3, const 0_isize);
```
Comment (clubby789) [inline tests/mir-opt/matches_reduce_branches.my_is_some.MatchBranchSimplification.diff:31]: Custom MIR limitations meant I couldn't have the branches in the same order, so this creates an Ne instead (unlike an Eq like the original demo)

**Comment by clubby789** [comment]
Just wanted to bump this and ask if this is desired/any changes are wanted

**Comment by DianQK** [comment]
r? DianQK

File: tests/mir-opt/matches_reduce_branches.rs [hunk 2]
```diff
@@ -19,6 +19,40 @@ fn foo(bar: Option<()>) {
     }
 }
 
+// EMIT_MIR matches_reduce_branches.my_is_some.MatchBranchSimplification.diff
+// Test for #131219. UnreachableEnumBranching introduces the third branch normally,
```
Comment (DianQK) [inline tests/mir-opt/matches_reduce_branches.rs:23]: Can you change the test case is:

```rust
pub fn my_is_some<T>(bar: Option<T>) -> bool {
    match bar {
        Some(_) => true,
        None => false,
    }
}
```

**Review by dianqk — APPROVED** [review]
LGTM. Thanks!
After modifying the test case, r=me.

File: tests/mir-opt/matches_reduce_branches.rs [hunk 3]
```diff
@@ -19,6 +19,18 @@ fn foo(bar: Option<()>) {
     }
 }
 
+// EMIT_MIR matches_reduce_branches.my_is_some.MatchBranchSimplification.diff
+// Test for #131219.
+fn my_is_some<T>(bar: Option<T>) -> bool {
```
Comment (DianQK) [inline tests/mir-opt/matches_reduce_branches.rs:24]:
```suggestion
fn my_is_some(bar: Option<()>) -> bool {
```

File: tests/mir-opt/matches_reduce_branches.rs [hunk 4]
```diff
@@ -651,4 +663,7 @@ fn main() {
     let _: u8 = match_trunc_u16_u8_failed(EnumAu16::u0_0x0000);
 
     let _ = match_i128_u128(EnumAi128::A);
+
+    let _ = my_is_some::<()>(None);
+    let _ = my_is_some(Some(()));
```
Comment (DianQK) [inline tests/mir-opt/matches_reduce_branches.rs:668]:
```suggestion
    let _ = my_is_some(None);
```

**Review by dianqk — COMMENTED** [review]
You cannot bless it again because of the different panic strategy. You can add `// EMIT_MIR_FOR_EACH_PANIC_STRATEGY`, but I prefer to keep simple here.

**Comment by clubby789** [comment]
```
  = note: LINK : fatal error LNK1104: cannot open file 'C:\a\rust\rust\build\x86_64-pc-windows-msvc\stage1-rustc\x86_64-pc-windows-msvc\release\deps\rustc_main-03a98b8d95adfc8a.exe'␍
```
@bors retry

**Comment by clubby789** [comment]
Hmm. This doesn't look related, but can't imagine what could cause it?

**Comment by clubby789** [comment]
@bors retry 
https://github.com/rust-lang/rust/issues/117097

**Pull request merged**